### ISSUE TITLE
Local settings in indent rule are not honored

### ISSUE THREAD

**Issue by pat-s** [cereda/arara#99]
https://github.com/cmhughes/latexindent.pl/issues/95

# Approach 1
If the reprex below is called with `arara` v4 and the file is reloaded, two `&` are falsely inserted.
These have to be manually removed before the next run otherwise it will fail.
This behaviour is accounted for in the `foo.yaml` that is specified in the `indent` rule in the header.
It is introduced by the special command `specialcell`.

Call: `arara foo.tex`

Although the `arara` call succeeds, the settings are not applied in the background.

# Approach 2
When calling the reprex with `arara` v4 and using the direct `latexindent.pl` script v3.3 afterwards, the local settings are honored and after file reload everything is fine.
When trying to replicate this approach the `indent` rule in the header has to be commented out (simply add another `%` before the rule).

Call: `arara foo.tex && latexindent -w -l foo.yaml foo.tex` 

# Reprex

```latex
%%%% using 'arara' 4.0
% arara: xelatex: {synctex: yes, interaction: nonstopmode}
% arara: indent: {overwrite: yes, settings: local, where: foo.yaml}
% arara: clean: { extensions: [ aux, bbl, bcf, cod, blg, lof, lot, out, toc, log, xml, bak0 ] }

\documentclass{article}

\usepackage{longtable}
\usepackage{booktabs}
\usepackage{multirow}
\usepackage{float}
\usepackage[export]{adjustbox}

% line breaks in table cells
\newcommand{\specialcell}[2][l]{%
  \begin{tabular}[#1]{@{}l@{}}#2\end{tabular}}

\begin{document}

\begin{table} [!t]
	\centering
	\begin{adjustbox}{max width=\textwidth}
		\begin{tabular}{*{3}{l|c|c|}}
			\specialcell{text \\ (text)}  & text & text \\
			\toprule
			text & text & text  \\
		\end{tabular}
	\end{adjustbox}
\end{table}

\end{document}
```

# foo.yaml

```
lookForAlignDelims:
   tabular:
      alignRowsWithoutMaxDelims: 0
      alignDoubleBackSlash: 0
```

**Comment by cereda** [comment]
Thanks, @pat-s! I will take a closer look later on!

**Comment by cereda** [comment]
Pat, I am so very, very sorry, but I am probably missing something very obvious. I tried the work flow here and the only difference I am getting is an extra space after `//` in `\specialcell`. I got no extra `&` symbols, so I don't know what's going on.

Both outputs from `arara` and `latexindent` issued directly gave me the exact same results. I am probably missing something here, and I deeply apologize. If it's not much trouble, could you post the output of each TeX file?

**Comment by pat-s** [comment]
:smile: that's odd Paulo! Wondering what the difference in our setups might be.

First, here is the misaligned output when using `arara foo.tex`.
As you can see it aligns the `\\` with the ones from the row below and therefore inserts two new fields that in the end cause the table having too many fields. Subsequently, a follow up build fails.

```latex
\begin{table} [!t]
	\centering
	\begin{adjustbox}{max width=\textwidth}
		\begin{tabular}{*{3}{l|c|c|}}
			\specialcell{text &      &      \\ (text)}  & text & text \\
			\toprule
			text              & text & text \\
		\end{tabular}
	\end{adjustbox}
\end{table}
```

I guess if you did not get a change or file reload after compiling, the indent rule did not execute at all on your side.

I just updated the header because there was still a different label for `foo.yaml` (`localSettings.yaml`). Maybe you didn't notice and thats why the rule was not applied. If so it was my fault and I am deeply sorry :(
Can you try again if it works now for you?

On the other hand, if you say that both outputs are the same and correct, that means that `foo.yaml` is honored on your side and we have to track down why this is not the case for me :confused: 

Can you reproduce the misaligned output by using `% arara: indent: {overwrite: yes}`?

**Comment by cereda** [comment]
Ouch, I don't know what is happening. I don't get your output in any way, either with or without `foo.yaml`. I am really, *really* lost. I first tried with `latexindent` shipped with my TL (properly updated), then I got the version from the tool repository. I still get the same output! It's insane.

I took the liberty of recording my session, see file below:

[My screencast (compacted, `.webm` format)](https://github.com/cereda/arara/files/1622604/screencast.zip)

Fedora 27, all updates, TL 2017, all updates. Help!

**Comment by cereda** [comment]
@cmhughes: Chris, if it's not much trouble, could you take a look at my screencast and confirm/deny that is the expected result?

**Comment by pat-s** [comment]
Paulo, don't worry. I get the same result as you when just using `latexindent`, even without local settings! Did not try this before though!

```
~/software/latexindent.pl-master/latexindent.pl ~/github/arara_reprex.tex

%%%% using 'arara' 4.0
% arara: xelatex: {synctex: yes, interaction: nonstopmode}
% arara: indent: {overwrite: yes, settings: local, where: localSettings.yaml}
% arara: clean: { extensions: [ aux, bbl, bcf, cod, blg, lof, lot, out, toc, log, xml, bak0 ] }

\documentclass{article}

\usepackage{longtable}
\usepackage{booktabs}
\usepackage{multirow}
\usepackage{float}
\usepackage[export]{adjustbox}

% line breaks in table cells
\newcommand{\specialcell}[2][l]{%
  \begin{tabular}[#1]{@{}l@{}}#2\end{tabular}}

\begin{document}

\begin{table} [!t]
	\centering
	\begin{adjustbox}{max width=\textwidth}
		\begin{tabular}{*{3}{l|c|c|}}
			\specialcell{text  \\ {text)}}  & text & text \\
			\toprule
			text & text & text \\
		\end{tabular}
	\end{adjustbox}
\end{table}

\end{document}
```
So there error really seems to be using arara AND the indent rule:

```
arara ~/github/arara_reprex.tex

\begin{table} [!t]
	\centering
	\begin{adjustbox}{max width=\textwidth}
		\begin{tabular}{*{3}{l|c|c|}}
			\specialcell{text &      &      \\ {text)}}  & text & text \\
			\toprule
			text              & text & text \\
		\end{tabular}
	\end{adjustbox}
\end{table}

\end{document}
```

Does the `arara`s indent rule grab the latexindent script from Tex Live? I am still having Tex Live 2015, just now installing Tex Live 2017 via backports and trying again.

**Comment by cereda** [comment]
Pat! This sounds like path priority issues. Let me try something here, I will report back in a couple of minutes.

**Comment by cereda** [comment]
Let's do this!

`which.yaml`

```yaml
!config
identifier: which
name: Which
commands:
- name: Carmen Sandiego
  command: >
    @{
        pair = unsafelyExecuteSystemCommand(getCommand('which', '-a', name));
        file = new java.io.File('path.txt');
        java.nio.file.Files.write(file.toPath(), pair.second().getBytes());
        return true;
    }
arguments:
- identifier: name
  flag: "@{parameters.name}"
  required: yes
```

`test.tex`:

```tex
% arara: which: { name: latexindent }
```

Execution:

```
  __ _ _ __ __ _ _ __ __ _ 
 / _` | '__/ _` | '__/ _` |
| (_| | | | (_| | | | (_| |
 \__,_|_|  \__,_|_|  \__,_|

Processing 'test.tex' (size: 39 bytes, last modified: 01/11/2018
11:58:10), please wait.

(Which) Carmen Sandiego ................................. SUCCESS

Total: 0.16 seconds
```

Then:

```
$ cat path.txt
/opt/texbin/latexindent
```

Let's see where is our `latexindent` being invoked!

**Comment by cereda** [comment]
Wow, `arara` 4.0 is a monster. I am getting legitimately scared!

**Comment by pat-s** [comment]
Paulo, after updating to the most recent Tex Live 2017 I don't get the error anymore :smile: Even without the custom YAML everything works!
So it really seemed to be caused by an outdated version of latexindent in Tex Live 2015?!

However, this still leaves the question open IF a local settings file would be honored :thinking: 
I guess the actual settings in the custom YAML file do no harm so we do not really see a difference in the outcome.

**Comment by cereda** [comment]
Patrick, I believe the execution will honour the file (I got some hints from my previous tests). If Chris could add a log entry, it would be very interesting to our cause! 😉 

Regarding TL, it seems the issue was related to path priority. In any case, the `which.yaml` rule posted might help us find out more about *which* binary/script is being executed at that specific moment.

I might come back to it at a later time, as I need to finish some drafts. Hopefully, the issue is solved for the time being. Let's keep an eye regarding the settings. 😉

**Comment by cmhughes** [comment]
Hi both,
Thanks for pinging me in this. I can verify that the output given in Paulo's screencast is as expected.

I can also verify that texlive 2015 will, necessarily, contain an outdated version of texlive. As we know, texlive freezes every year, so texlive 2015 will be very out of date. I released version 3.0 of latexindent in Feb 2017, so it would have been a part of texlive 2016 and texlive 2017.

With regards to the log file, when I run
```
latexindent.pl myfile.tex
```
then I get the following `indent.log`
```
latexindent.pl version 3.3, 2017-08-21, a script to indent .tex files
	latexindent.pl lives here: /home/cmhughes/projects/latexindent/
	Thu Jan 11 21:00:21 2018
Processing switches
Directory for backup files and indent.log: .
Perl modules are being loaded from the following directories:
	/home/cmhughes/perl5/perlbrew/perls/perl-5.26.0/lib/5.26.0/FindBin.pm
	/home/cmhughes/perl5/perlbrew/perls/perl-5.26.0/lib/site_perl/5.26.0/YAML/Tiny.pm
	/home/cmhughes/perl5/perlbrew/perls/perl-5.26.0/lib/5.26.0/File/Copy.pm
	/home/cmhughes/perl5/perlbrew/perls/perl-5.26.0/lib/5.26.0/File/Basename.pm
	/home/cmhughes/perl5/perlbrew/perls/perl-5.26.0/lib/5.26.0/Getopt/Long.pm
	/home/cmhughes/perl5/perlbrew/perls/perl-5.26.0/lib/site_perl/5.26.0/File/HomeDir.pm
	/home/cmhughes/perl5/perlbrew/perls/perl-5.26.0/lib/site_perl/5.26.0/i686-linux/Unicode/GCString.pm
Latex Indent perl modules are being loaded from, for example:
	/home/cmhughes/projects/latexindent/LatexIndent/Document.pm
YAML settings read: defaultSettings.yaml
	Reading defaultSettings.yaml from /home/cmhughes/projects/latexindent/defaultSettings.yaml
YAML settings read: indentconfig.yaml or .indentconfig.yaml
	Reading path information from /home/cmhughes/.indentconfig.yaml
	(Alternatively /home/cmhughes/indentconfig.yaml can be used)
	---
	paths:
	  - /home/cmhughes/Documents/yamlfiles/mysettings.yaml

YAML settings, reading from the following files:
	Reading USER settings from /home/cmhughes/Documents/yamlfiles/mysettings.yaml
	Not showing settings in the log file (see showEveryYamlRead and showAmalgamatedSettings).
Phase 1: searching for objects
Phase 2: finding surrounding indentation
Phase 3: indenting objects
Phase 4: final indentation check
Output routine
	Not outputting to file; see -w and -o switches for more options.
--------------
Please direct all communication/issues to: 
	https://github.com/cmhughes/latexindent.pl
```
when I run
```
latexindent.pl pat.tex -l=foo.yaml
```
I get the following `indent.log`
```
latexindent.pl version 3.3, 2017-08-21, a script to indent .tex files
	latexindent.pl lives here: /home/cmhughes/projects/latexindent/
	Thu Jan 11 21:00:56 2018
Processing switches
	-l|--localSettings: Read localSettings YAML file
Directory for backup files and indent.log: .
Perl modules are being loaded from the following directories:
	/home/cmhughes/perl5/perlbrew/perls/perl-5.26.0/lib/5.26.0/FindBin.pm
	/home/cmhughes/perl5/perlbrew/perls/perl-5.26.0/lib/site_perl/5.26.0/YAML/Tiny.pm
	/home/cmhughes/perl5/perlbrew/perls/perl-5.26.0/lib/5.26.0/File/Copy.pm
	/home/cmhughes/perl5/perlbrew/perls/perl-5.26.0/lib/5.26.0/File/Basename.pm
	/home/cmhughes/perl5/perlbrew/perls/perl-5.26.0/lib/5.26.0/Getopt/Long.pm
	/home/cmhughes/perl5/perlbrew/perls/perl-5.26.0/lib/site_perl/5.26.0/File/HomeDir.pm
	/home/cmhughes/perl5/perlbrew/perls/perl-5.26.0/lib/site_perl/5.26.0/i686-linux/Unicode/GCString.pm
Latex Indent perl modules are being loaded from, for example:
	/home/cmhughes/projects/latexindent/LatexIndent/Document.pm
YAML settings read: defaultSettings.yaml
	Reading defaultSettings.yaml from /home/cmhughes/projects/latexindent/defaultSettings.yaml
YAML settings read: indentconfig.yaml or .indentconfig.yaml
	Reading path information from /home/cmhughes/.indentconfig.yaml
	(Alternatively /home/cmhughes/indentconfig.yaml can be used)
	---
	paths:
	  - /home/cmhughes/Documents/yamlfiles/mysettings.yaml

YAML settings read: -l switch
	Adding foo.yaml to YAML read paths
YAML settings, reading from the following files:
	Reading USER settings from /home/cmhughes/Documents/yamlfiles/mysettings.yaml
	Not showing settings in the log file (see showEveryYamlRead and showAmalgamatedSettings).
	Reading USER settings from foo.yaml
	Not showing settings in the log file (see showEveryYamlRead and showAmalgamatedSettings).
Phase 1: searching for objects
Phase 2: finding surrounding indentation
Phase 3: indenting objects
Phase 4: final indentation check
Output routine
	Not outputting to file; see -w and -o switches for more options.
--------------
Please direct all communication/issues to: 
	https://github.com/cmhughes/latexindent.pl
```

Notice, in particular, that the settings from the `-l` switch are detailed. 

Is this as you both need it? It looks like things are working as expected at my end, but do let me know if not :)

**Comment by cereda** [comment]
Chris, thank you very much, and sorry for not noticing the entry before, I should've spot it...

Invalid file:

```tex
% arara: indent: { settings: local, where: nofile.yaml }
```

Output:

```
YAML settings read: -l switch
	WARNING yaml file not found: nofile.yaml not found. Proceeding without it.
```

Valid file:

```tex
% arara: indent: { settings: local, where: foo.yaml }
```

Output:

```
YAML settings read: -l switch
	Adding foo.yaml to YAML read paths
```

Patrick, I think the mystery is solved. 😉

**Comment by pat-s** [comment]
Ok :tada:

So to summarize:

I needed to update my Tex Live and in Version 2017 the "fix" was already available without the need of the custom yaml file.

The log output shows if a custom yaml file is used or not :smile: 

Please correct me if I got something wrong! And a HUGE thanks to you both :smile: :duck: !

**Comment by cmhughes** [comment]
Great! So glad this is resolved.

In V3.4 of latexindent the warning will be a bit more obvious, as the
logging method is enhanced.

**Comment by cmhughes** [comment]
I meant to say that you can check your version of `latexindent.pl` by typing
```
latexindent.pl -v
```
or

```
latexindent.pl -version
```

As a side note, I released V3.4 to ctan today :)

**Comment by cereda** [comment]
Huzzah! Congrats, Chris!
